custom_mutators/rust/custom_mutator/src/lib.rs

```diff
@@ -0,0 +1,490 @@
+//! Somewhat safe and somewhat ergonomic bindings for creating [AFL++](https://github.com/AFLplusplus/AFLplusplus) [custom mutators](https://github.com/AFLplusplus/AFLplusplus/blob/stable/docs/custom_mutators.md) in Rust.
+//!
+//! # Usage
+//! AFL++ custom mutators are expected to be dynamic libraries which expose a set of symbols.
+//! Check out [`CustomMutator`] to see which functions of the API are supported.
+//! Then use [`export_mutator`] to export the correct symbols for your mutator.
+//! In order to use the mutator, your crate needs to be a library crate and have a `crate-type` of `cdylib`.
+//! Putting
+//! ```yaml
+//! [lib]
+//! crate-type = ["cdylib"]
+//! ```
+//! into your `Cargo.toml` should do the trick.
+//! The final executable can be found in `target/(debug|release)/your_crate_name.so`.
+//! # Example
+//! See [`export_mutator`] for an example.
+//!
+//! # On `panic`s
+//! This binding is panic-safe in that it will prevent panics from unwinding into AFL++. Any panic will `abort` at the boundary between the custom mutator and AFL++.
+//! 
+//! # Access to AFL++ internals
+//! This crate has an optional feature "afl_internals", which gives access to AFL++'s internal state.
+//! The state is passed to [`CustomMutator::init`], when the feature is activated.
+//! 
+//! _This is completely unsafe and uses automatically generated types extracted from the AFL++ source._
+pub mod fallible;
+
+use std::{ffi::CStr, os::raw::c_uint};
+
+#[cfg(feature = "afl_internals")]
+#[doc(hidden)]
+pub use custom_mutator_sys::afl_state;
+
+/// The result of a call to [`CustomMutator::fuzz`].
+#[derive(Debug)]
+pub enum FuzzResult<'l> {
+    /// Returned when the given buffer was modified in place. Modifying the buffer in place to be of a different length is not supported currently.
+    InPlace,
+    /// Returned when a buffer that is managed by the custom mutator should be returned.
+    NewBuffer(&'l [u8]),
+    /// Returned when the custom mutator fails. This will make AFL++ fail fast (ie crash) and corresponds to returning a NULL buffer and 0 length.
```

Comment:
Not sure if this is true for all functions, I remember there was a PR to allow 0 sized returns for custom mutators, they will be skipped then (do you remember, @vanhauser-thc?)

Reply:
cant remember would have to check the source too. if it is not documented in docs/ it should be added

Reply:
AFAICT:
* 0 sized return will skip the mutation
* null buffer return will `FATAL` in AFL++.

I think a FATAL in AFL++ should never be the goals from inside a custom mutator. if a custom mutator needs to abort, it can do so itself (ie. by `panic`ing). 
In my latest refactoring ([comment below](https://github.com/AFLplusplus/AFLplusplus/pull/752#issuecomment-781956576)), the behaviour would be to allow the custom mutator to return a 0 size to skip the mutation, but there would be no way to return a null pointer.